src/include/daos_api.h

```diff
@@ -220,11 +220,149 @@ daos_prop_alloc(uint32_t entries_nr);
+ */
+struct daos_ace *
+daos_acl_get_first_ace(struct daos_acl *acl);
+
+/**
+ * Get the next Access Control Entry in the Access Control List, for iterating
+ * over the list.
+ *
+ * \param[in]	acl		ACL to traverse
+ * \param[in]	current_ace	Current ACE, to determine the next one
+ *
+ * \return	Pointer to the next ACE in the ACL, or NULL if at the end
+ */
+struct daos_ace *
+daos_acl_get_next_ace(struct daos_acl *acl, struct daos_ace *current_ace);
+
+/**
+ * Search the Access Control List for an Access Control Entry for a specific
+ * principal.
+ *
+ * \param[in]	acl		ACL to search
+ * \param[in]	type		Principal type to search for
+ * \param[in]	principal	Principal name, if type is USER or GROUP. NULL
+ *				otherwise.
+ *
+ * \return	Pointer to the matching ACE, or NULL if not found
+ */
+struct daos_ace *
+daos_acl_get_ace_for_principal(struct daos_acl *acl,
+		enum daos_acl_principal_type type, const char *principal);
+
+/**
+ * Insert an Access Control Entry in the appropriate location in the ACE
+ * list. The expected order is: Owner, Users, Assigned Group, Groups, Everyone.
+ *
+ * This requires us to reconstruct and reallocate the ACL structure. Both the
+ * old structure and the new one must be freed by the caller. The old data is
+ * left unchanged.
+ *
+ * If the new ACE is an update of an existing entry, it will replace the old
+ * entry.
+ *
+ * \param[in]	acl	Original ACL
+ * \param[in]	new_ace	ACE to be added
+ * \param[out]	new_acl	Reallocated copy of the ACL with new item inserted
+ *
+ * \return	DER_SUCCESS or error code
+ */
+int
+daos_acl_add_ace_realloc(struct daos_acl *acl, struct daos_ace *new_ace,
+		struct daos_acl **new_acl);
+
+/**
+ * Remove an Access Control Entry from the list.
+ *
+ * This requires us to reconstruct and reallocate the ACL structure. Both the
+ * old structure and the new one must be freed by the caller. The old data is
+ * left unchanged.
+ *
+ * \param[in]	acl			Original ACL
+ * \param[in]	type			Principal type of the ACE to remove
+ * \param[in]	principal_name		Principal name of the ACE to remove
+ *					(NULL if type isn't user/group)
+ * \param[in]	principal_name_len	Length of the principal_name string
+ * \param[out]	new_acl			Reallocated copy of the ACL with the
+ *					ACE removed
+ *
+ * \return	DER_SUCCESS or error code
+ */
+int
+daos_acl_remove_ace_realloc(struct daos_acl *acl,
+		enum daos_acl_principal_type type, const char *principal_name,
+		size_t principal_name_len, struct daos_acl **new_acl);
+
+/**
+ * Allocate a new Access Control Entry with an appropriately aligned principal
+ * name, if applicable.
+ *
+ * Only User and Group types use principal name.
+ *
+ * \param[in]	type			Type of principal for the ACE
+ * \param[in]	principal_name		Principal name will be added to the end
+ *					of the structure. For types that don't
+ *					use it, it is ignored. OK to pass NULL.
+ * \param[in]	principal_name_len	Length of the principal_name string
```

Comment:
why do we need the name len? can you just say it has to be NULL terminated?

Reply:
Just another safety rail against buffer overruns. If it's safe for us to assume the input here is always null terminated, I can do that.

Reply:
I would keep the name length since this will be on disk and should be there as a sanity check to ensure what's on disk makes sense.

Reply:
@dpquigl We do still align up the length to 64 bits within the function itself. We'd need to derive the minimum length from `strlen(principal_name)` within the function if the caller no longer passes it in - but if we don't need to be super defensive with client applications, maybe that's fine. And of course they could lie about the length anyway.

Reply:
If called by external code, we can do little about string (and pointer) arguments. It's callers' memory anyway. If called by other parts of libdaos, but still with external-code-supplied strings (or pointers), the same. If called by daos_io_server, we shall carefully execute those considerations above _before_ calling functions like this. Just a thought for your consideration.

Reply:
@liw That's a good argument. I have been leaning toward that conclusion myself.